contracts/d3vent.sol

```diff
@@ -0,0 +1,165 @@
+    // should joiner address always be msg.sender or allow to join another address?
+    
+    // don't think there's any need to zero out joiner balances after organiser withdrawal, but could add a flag for added safety
+    // would be too gassy iterating through a large number of joiners
+    // 
+    
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.7;
+
+contract d3vent {
+    event CreatedEvent(uint indexed eventId, uint indexed eventDate, string indexed eventName);
+    event Withdrawal(uint indexed eventId, address indexed organiser, uint balance);
+    event JoinableSet(uint indexed eventId, bool isJoinable);
+       
+    uint eventIds;
+    uint immutable withdrawalBuffer;
+
+    //uint immutable unjoinBuffer; // not implemented
+    
+    // not implemented
+    // struct user {
+    //     address userAddress;
+    //     event_[] eventIds;  // do we need quick lookups and/or able to iterate eventIds?
+    // }
+
+    struct event_ {
+        address organiser;
+        uint id;
+        string name;
+        string uri;
+        uint dateTime;
+        uint price;
+        uint128 capacity;
+        uint128 numJoined;      
+        bool isJoinable;
+        uint withdrawalDate;
+    }
+
+    event_[] events;
+
+
+    mapping(uint => mapping(address => bool)) isJoined;    // eventId => userAddr => isJoined
+    mapping(uint => mapping(address => uint)) joinerBalances;   // eventId => userAddr => balance
+    mapping(uint => uint) eventBalances;    // eventId => event balance
+    mapping(address => bool) isAdmin;   // admin address to bool
+    
+
+    constructor (uint _withdrawalBuffer) {
+        withdrawalBuffer = _withdrawalBuffer;
+    }
+    
+    modifier isAuthorised (uint _eventId) {
+        require(
+            msg.sender == events[_eventId].organiser || 
+            isAdmin[msg.sender], 
+            "not authroised"
+        );
+        _;
+    }
+
+    // test parameters: "test event name","https://livepeer.org/123",1662994265,1000,1000000,false
+    function createEvent(
+        //string calldata _eventName,
+        string calldata _name,
+        string calldata _uri,
+        uint _dateTime,
+        uint _price,
+        uint128 _capacity,
+        bool _isJoinable
+        ) external {              // public to allow internal calls for testing
+
+        // sanity checks
+        require(_dateTime > block.timestamp, "date/time in past");
+
+        event_ memory newEvent;
+        
+        newEvent.organiser = msg.sender;
+        newEvent.id = eventIds++;
+        newEvent.name = _name;
+        newEvent.uri = _uri;
+        newEvent.dateTime = _dateTime;
+        newEvent.price = _price;
+        newEvent.capacity = _capacity;
+        newEvent.isJoinable = _isJoinable;
+    
+        events.push(newEvent);
+        emit CreatedEvent(newEvent.id, newEvent.dateTime, newEvent.name);
+    }
+
+
+    function setOrganiser(uint _eventId, address _newOrganiser) external {
+        // sanity checks
```

Comment:
setOrganiser function should only be called by the existing organizer of that particular event. Otherwise it can generate vulnerabilities.

Reply:
yep. well spotted need to add organiser checks on a few functions

Reply:
👍